torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,881 @@ namespace fuser {
+    projectIdToRFactor(reference, reordered_root);
   }
 }
 
-// Merge innermost domains for finding the widest vectorizable
-// size. Return the merged domain or nullptr if no merge is done.
-IterDomain* mergeInnermostDomains(
-    const std::vector<IterDomain*>& domain,
-    int num_merged_domains) {
-  const auto ndims = domain.size();
-  IterDomain* merged_id = nullptr;
-  bool is_merge_done = false;
-  for (const auto i : c10::irange(num_merged_domains)) {
-    auto id = domain.at(ndims - 1 - i);
-    // broadcast and trivial reductions are ignored
-    if (id->isBroadcast() || id->isTrivialReduction()) {
-      continue;
-    }
-    if (merged_id == nullptr) {
-      merged_id = id;
+std::unordered_map<TensorView*, std::vector<IterDomain*>>::iterator
+ContiguousInnerDimensionsMapper::projectIdToRoot(
+    TensorView* ref,
+    std::vector<IterDomain*> ids) {
+  auto transform_exprs = StmtSort::getExprs(
+      ref->fusion(),
+      {ref->getRFactorDomain().begin(), ref->getRFactorDomain().end()});
+
+  // Mapping from rfactor to root, reverse expressions
+  std::reverse(transform_exprs.begin(), transform_exprs.end());
+
+  for (const auto* expr : transform_exprs) {
+    if (const Split* split = dynamic_cast<const Split*>(expr)) {
+      // Initialize state
+      auto find_outer_it = ids.begin();
+      auto outer_pos = ids.size();
+      auto find_inner_it = ids.begin();
+      auto inner_pos = ids.size();
+
+      // Removes all entries to the left of provided `it`, if `it` is not
+      // ids.begin(). Updates all state of finding outer and inner in the ids
+      // vector after erasing.
+      auto clear_left_of = [&find_outer_it,
+                            &outer_pos,
+                            &find_inner_it,
+                            &inner_pos,
+                            &ids,
+                            &split](decltype(find_outer_it) it) {
+        if (it != ids.begin()) {
+          ids.erase(ids.begin(), it);
+        }
+
+        // Set outer it and position
+        find_outer_it = std::find(ids.begin(), ids.end(), split->outer());
+        outer_pos = find_outer_it == ids.end()
+            ? ids.size()
+            : std::distance(ids.begin(), find_outer_it);
+
+        // Set inner it and position
+        find_inner_it = std::find(ids.begin(), ids.end(), split->inner());
+        inner_pos = find_inner_it == ids.end()
+            ? ids.size()
+            : std::distance(ids.begin(), find_inner_it);
+      };
+
+      // Dry run to fill state
+      clear_left_of(ids.begin());
+
+      // Check if the domains out of the split are contiguous in the mapped
+      // domain.
+      if (find_outer_it == ids.end() && find_inner_it != ids.end()) {
+        // Outer dimension was not found, but inner dimension was. Must assume
+        // everything to the left of inner is not contiguously merged.
+        //
+        // Clear left of inner
+        clear_left_of(find_inner_it);
```

Comment:
Will add a comment. We want to continue processing the statement here because inner maps so we might have a partial inner mapping to continue with. We just don't have outer so anything left of inner can't partially map.

Reply:
Comment added will be:
```
      // There are a few continue conditions below that stop processing the
      // expression. Continue is used when we cannot map any information even
      // partial of the expression.
      //
      // If inner is unmapped.
      // If neither inner or outer are mapped.
      // If inner and outer are mapped but inner is mapped outside of outer.
      //
      // If the above conditions aren't hit, then we want to continue
      // propagation of a partial mapping through the expressions.
      ```